torch/nn/functional.py

```diff
@@ -1003,72 +1003,53 @@ def softmax(input, dim=None, _stacklevel=3, dtype=None):
         return input.softmax(dim, dtype=dtype)
 
 
-def _sample_gumbel(shape, eps=1e-10, out=None):
-    """
-    Sample from Gumbel(0, 1)
-
-    based on
-    https://github.com/ericjang/gumbel-softmax/blob/3c8584924603869e90ca74ac20a6a03d99a91ef9/Categorical%20VAE.ipynb ,
-    (MIT license)
-    """
-    U = out.resize_(shape).uniform_() if out is not None else torch.rand(shape)
-    return - torch.log(eps - torch.log(U + eps))
-
-
-def _gumbel_softmax_sample(logits, tau=1, eps=1e-10):
-    """
-    Draw a sample from the Gumbel-Softmax distribution
-
-    based on
-    https://github.com/ericjang/gumbel-softmax/blob/3c8584924603869e90ca74ac20a6a03d99a91ef9/Categorical%20VAE.ipynb
-    (MIT license)
-    """
-    dims = logits.dim()
-    gumbel_noise = _sample_gumbel(logits.size(), eps=eps, out=logits.data.new())
-    y = logits + gumbel_noise
-    return softmax(y / tau, dims - 1)
-
-
-def gumbel_softmax(logits, tau=1, hard=False, eps=1e-10):
+def gumbel_softmax(logits, tau=1, hard=False, eps=1e-10, dim=-1):
     r"""
-    Sample from the Gumbel-Softmax distribution and optionally discretize.
+    Samples from the `Gumbel-Softmax distribution`_ and optionally discretizes.
 
     Args:
-      logits: `[batch_size, num_features]` unnormalized log probabilities
+      logits: `[..., num_features]` unnormalized log probabilities
       tau: non-negative scalar temperature
       hard: if ``True``, the returned samples will be discretized as one-hot vectors,
             but will be differentiated as if it is the soft sample in autograd
+      eps: Deprecated parameter to avoid log(0). Now handled elsewhere.
+      dim (int): A dimension along which softmax will be computed.
```

Comment:
Maybe would be better to elide eps from the docs, and specify that dims is keyword only.

Reply:
dims should say what the default is; as spelled right now it implies the argument is mandatory (which it is not)

Reply:
Fixed with a02c8b9 and ea1b998